app/uk/gov/hmrc/exports/services/reversemapping/declaration/items/ProcedureCodesParser.scala

```diff
@@ -0,0 +1,74 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.services.reversemapping.declaration.items
+
+import uk.gov.hmrc.exports.models.StringOption
+import uk.gov.hmrc.exports.models.declaration.ProcedureCodes
+import uk.gov.hmrc.exports.services.reversemapping.declaration.DeclarationXmlParser.XmlParserResult
+import uk.gov.hmrc.exports.services.reversemapping.declaration.XmlTags._
+import uk.gov.hmrc.exports.services.reversemapping.declaration.{DeclarationXmlParser, XmlParsingException}
+
+import scala.xml.NodeSeq
+
+class ProcedureCodesParser extends DeclarationXmlParser[Option[ProcedureCodes]] {
+
+  override def parse(itemXml: NodeSeq): XmlParserResult[Option[ProcedureCodes]] = validateInput(itemXml).map { _ =>
+    val procedureCode: Option[String] = (itemXml \ GovernmentProcedure).find { governmentProcedureNode =>
+      (governmentProcedureNode \ PreviousCode).nonEmpty
+    }.flatMap { governmentProcedureNode =>
+      for {
+        currentCode <- StringOption((governmentProcedureNode \ CurrentCode).text)
+        previousCode <- StringOption((governmentProcedureNode \ PreviousCode).text)
+      } yield currentCode + previousCode
+    }
+
+    val additionalProcedureCodes: Seq[String] = (itemXml \ GovernmentProcedure).flatMap { governmentProcedureNode =>
+      governmentProcedureNode
+        .filterNot(node => StringOption((node \ PreviousCode).text).nonEmpty)
+        .flatMap { governmentProcedure =>
+          StringOption((governmentProcedure \ CurrentCode).text)
+        }
+    }
+
+    if (procedureCode.isDefined || additionalProcedureCodes.nonEmpty)
+      Some(ProcedureCodes(procedureCode, additionalProcedureCodes))
+    else None
+
+  }
+
+  private def validateInput(itemXml: NodeSeq): XmlParserResult[NodeSeq] = {
+    val isThereGovernmentProcedureWithPreviousCodeOnly = (xml: NodeSeq) =>
+      (xml \ GovernmentProcedure).exists { governmentProcedureNode =>
+        (governmentProcedureNode \ PreviousCode).nonEmpty && (governmentProcedureNode \ CurrentCode).isEmpty
+    }
+
+    val areThereMultipleGovernmentProceduresWithPreviousCodes = (xml: NodeSeq) =>
+      (xml \ GovernmentProcedure).count { governmentProcedureNode =>
+        (governmentProcedureNode \ PreviousCode).nonEmpty
+      } > 1
+
+    val errorMessages = Map(
+      isThereGovernmentProcedureWithPreviousCodeOnly -> "GovernmentProcedure element does not contain CurrentCode element",
+      areThereMultipleGovernmentProceduresWithPreviousCodes -> "There are multiple GovernmentProcedure elements with PreviousCode elements"
+    )
+
+    val errors = errorMessages.filter { case (validationFunction, _) => validationFunction(itemXml) }.values
+
+    if (errors.nonEmpty) Left(XmlParsingException(errors.mkString("[", ", ", "}")))
+    else Right(itemXml)
+  }
```

Comment:
Just wondering where this logic is defined (is it in the tickets Beth and Monica wrote or is this just something you knew from past experience)?

Reply:
This should be present in the tickets, but you can always have a look into wco builders. In this case it's [TypeCodeBuilder](https://github.com/hmrc/customs-declare-exports/blob/61e25a825e5fbf0c04a8805a7ecee0a0fa33d0f2/app/uk/gov/hmrc/exports/services/mapping/declaration/TypeCodeBuilder.scala). And personally, I prefer to look into the code, because this is how it really works, not some spreadsheet extract.